### ISSUE TITLE
Authentication using flat databases

### ISSUE THREAD

**Issue by foxcpp** [foxcpp/maddy#99]
Currently, the following authentication providers are supported by maddy:
* go-imap-sql native authentication
  Works best when go-imap-sql is also used for storage, brings a lot of unnecessary things when used only for authentication.
* PAM-based authentication
  Intended for use with PAM-based system account authentication. Doesn't support other use-cases.
* shadow-based authentication
  Intended for use with Unix system accounts. Reasonably simple and does only what is needed, but also suffers from legacy hashing algorithms.

When server admins just want "virtual users" with a custom database for accounts. Probably using the potential maildir backend discussed in #6. What should they use for authentication? Perhaps we need to have another auth. a provider using some dumb file format like USERNAME:HASHALGO:SALT:PASSHASH to do just that.

**Comment by emersion** [comment]
Most people will probably prefer PAM or shadow for this use-case, I think.

**Comment by foxcpp** [comment]
Side note: Both providers don't actually support auth_perdomain because they don't allow @ to be used as part of the login name.

Side note: Probably I should make a PR to reject the following configuration as invalid, currently the first two directives will be silently ignored.
```
auth_perdomain
auth_domains example.org example.com

imap ... {
  auth pam { }
}
```

**Comment by foxcpp** [comment]
>Both providers don't actually support auth_perdomain because they don't allow @ to be used as part of the login name.

So that leaves us without the possibility to create an isolated namespace in a system DB. Meaning that if you want `foxcpp` account, you can't have different passwords for SSH and e-mail.

**Comment by foxcpp** [comment]
Related: #56.

**Comment by foxcpp** [comment]
Feedback from IRC (and also the reason of removal of "rfc" label, "we need it"):

> < jpsamaroo > +1 to auth in a DB other than shadow or PAM. i don't want to have to create system users just to have additional email accounts   

> < lupine > I want my email users to be completely separate to my system users. it's quite a common scenario

**Comment by MFAshby** [comment]
> a provider using some dumb file format like USERNAME:HASHALGO:SALT:PASSHASH to do just that.

I think that [Apache htpasswd](https://httpd.apache.org/docs/current/programs/htpasswd.html) produces files in this format already if it helps.

**Comment by foxcpp** [comment]
>I think that Apache htpasswd produces files in this format already if it helps.

I think attempting to stay compatible with any existing format will bring a set of undesired restrictions. Since we have `maddyctl users ...` command anyway that is able to transparently handle any authentication backend used by maddy, there is no point in compatibility either.

**Comment by foxcpp** [comment]
Implemented in cdd01c8e37014cd41d7d2f54154f9da468551ec5. Might need some improvements but it works.

```
auth pass_table file_table /etc/maddy/passwd
```
